AXI Memory Over TL Serial Link (#812)
ucb-bar/chipyard: docs/Advanced-Concepts/Chip-Communication.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, Chipyard uses the Tethered Serial Interface (TSI) to communicate with the DUT.
TSI protocol is an implementation of HTIF that is used to send commands to the RISC-V DUT.

> These TSI commands are simple R/W commands that are able to probe the DUT's memory space.

Reviewer: probe->access

During simulation, the host sends TSI commands to a simulation stub in the test harness called ``SimSerial``
(C++ class) that resides in a ``SimSerial`` Verilog module (both are located in the ``generators/testchipip``
project).
This ``SimSerial`` Verilog module then sends the TSI command recieved by the simulation stub
to an adapter that converts the TSI command into a TileLink request.
This conversion is done by the ``SerialAdapter`` module (located in the ``generators/testchipip`` project).
After the transaction is converted to TileLink, the ``TLSerdesser`` (located in ``generators/testchipip``) serializes the
transaction and sends it to the chip (this ``TLSerdesser`` is sometimes also referred to as a digital serial-link or SerDes).
Once the serialized transaction is received on the chip, it is deserialized and masters a TileLink bus on the chip
which handles the request.
In simulation, FESVR resets the DUT, writes into memory the test program, and indicates to the DUT to start the program
through an interrupt (see :ref:`customization/Boot-Process:Chipyard Boot Process`).

> Using TSI is currently the fastest mechanism to communicate with the DUT in simulation and is also used by FireSim.

Reviewer: If this is in comparison to DTM/JTAG it might be worth mentioning that? Also not sure if it is worth talking about the width here.

Author: I think we can glide over the width here since that is really an implementation detail that can be changed (if the user understands what TSI is). In any case we briefly discuss AXI vs TSI widths in the testchip section.

This simulation stub creates a socket that OpenOCD and GDB can connect to when the simulation is running.
The default Chipyard designs instantiate the DTM configured to use JTAG (i.e. ``RocketConfig``).

.. note::
    As mentioned, default Chipyard designs are enabled with JTAG.
    However, they also use TSI/Serialized-TL with FESVR in case the JTAG interface isn't used.
    This allows users to choose how to communicate with the DUT (use TSI or JTAG).

Debugging with JTAG
~~~~~~~~~~~~~~~~~~~

Roughly the steps to debug with JTAG in simulation are as follows:

1. Build a Chipyard JTAG-enabled RTL design. Remember default Chipyard designs are JTAG ready.

.. code-block:: bash

    cd sims/verilator
    # or
    cd sims/vcs

    make CONFIG=RocketConfig

2. Run the simulation with remote bit-bang enabled. Since we hope to load/run the binary using JTAG,
   we can pass ``none`` as a binary (prevents FESVR from loading the program). (Adapted from: https://github.com/chipsalliance/rocket-chip#3-launch-the-emulator)

.. code-block:: bash

    # note: this uses Chipyard make invocation to run the simulation to properly wrap the simulation args

>     make CONFIG=RocketConfig BINARY=none SIM_FLAGS="+jtag_rbb_enable=1 --rbb-port=9823" run-binary

Reviewer: Great!

3. `Follow the instructions here to connect to the simulation using OpenOCD + GDB. <https://github.com/chipsalliance/rocket-chip#4-launch-openocd>`__

.. note::
    This section was adapted from the instruction in Rocket Chip and riscv-isa-sim. For more information refer
    to that documentation: `Rocket Chip GDB Docs <https://github.com/chipsalliance/rocket-chip#-debugging-with-gdb>`__,
    `riscv-isa-sim GDB Docs <https://github.com/riscv/riscv-isa-sim#debugging-with-gdb>`__

Example Test Chip Bringup Communication
---------------------------------------

Intro to Typical Chipyard Test Chip
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Most, if not all, Chipyard configurations are tethered using TSI (over a serial-link) and have access

Reviewer: I'd like this sentence to make it clear that having TSI doesn't mean all operation will be tethered, but I don't know how to do that.

Author: I think this is fine since this entire section is referring to "things that inspect the target memory for testing purposes". Its not talking about different ports off the DUT, etc. (If I'm understanding your question correctly).

to external memory through an AXI port (backing AXI memory).
The following image shows the DUT with these set of default signals:

.. image:: ../_static/images/default-chipyard-config-communication.png

In this setup, the serial-link is connected to the TSI/FESVR peripherals while the AXI port is connected
to a simulated AXI memory.

> However, AXI ports tend to have many signals associated with them so instead of creating an AXI port off the DUT,

Reviewer: "many signals, and thus wires,"

serial-link.

.. image:: ../_static/images/chip-bringup-simulation.png

.. note::
    Here the simulated AXI memory and the converters can be in a different clock domain in the test harness
    than the reference clock of the DUT.
    For example, the DUT can be clocked at 3.2GHz while the simulated AXI memory can be clocked at 1GHz.
    This functionality is done in the harness binder that instantiates the TSI collateral, TL-to-AXI converters,
    and simulated AXI memory.
    See :ref:`Advanced-Concepts/Harness-Clocks:Creating Clocks in the Test Harness` on how to generate a clock
    in a harness binder.

This type of simulation setup is done in the following multi-clock configuration:

.. literalinclude:: ../../generators/chipyard/src/main/scala/config/RocketConfigs.scala
    :language: scala
    :start-after: DOC include start: MulticlockAXIOverSerialConfig
    :end-before: DOC include end: MulticlockAXIOverSerialConfig

Bringup Setup of the Example Test Chip after Tapeout
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Assuming this example test chip is taped out and now ready to be tested, we can communicate with the chip using this serial-link.

> For example, some Berkeley tapeouts of Chipyard chips have an FPGA running a RISC-V soft-core that is able to speak to the DUT.

Reviewer 2: ```suggestion
For example, a common test setup used at Berkeley to evaluate Chipyard-based test-chips includes an FPGA running a RISC-V soft-core that is able to speak to the DUT (over an FMC).
```

This RISC-V soft-core would serve as the host of the test that will run on the DUT.
This is done by the RISC-V soft-core running FESVR, sending TSI commands to a ``SerialAdapter`` / ``TLSerdesser`` programmed on the FPGA.
Once the commands are converted to serialized TileLink, then they can be sent over some medium to the DUT
(like an FMC cable or a set of wires connecting FPGA outputs to the DUT board).
Similar to simulation, if the chip requests offchip memory, it can then send the transaction back over the serial-link.

> Then the request can be serviced by the channel of FPGA DRAM.

Reviewer: ```suggestion
Then the request can be serviced by the FPGA DRAM.
```

------------------------

The main way to use JTAG with a Rocket Chip based system is to instantiate the Debug Transfer Module (DTM)
and configure it to use a JTAG interface.
The default Chipyard designs instantiate the DTM and configure it to use JTAG.
You may attach OpenOCD and GDB to any of the default JTAG-enabled designs.

Debugging with JTAG
~~~~~~~~~~~~~~~~~~~

Please refer to the following resources on how to debug a Rocket Chip system with JTAG.

* https://github.com/chipsalliance/rocket-chip#-debugging-with-gdb

> * https://github.com/riscv/riscv-isa-sim#debugging-with-gdb

Reviewer 2: Should we move these lines to be after step 2, so people don't jump directly to rocketchip and try to go by the full rocketchip instructions, but rather start by building the chipyard configs and running the chipyard simulation, and then using the rocketchip instructions only for OpenOCD and GDB instructions?

Author: I reworked this a bit. Check now.

Roughly the steps are as follows (refer to the links for details):

1. Build a Chipyard JTAG-enabled RTL design (i.e. ``make CONFIG=RocketConfig`` in ``sims/*``)
2. Run the simulation with remote bit-bang enabled (i.e. ``make CONFIG=RocketConfig BINARY=<YourBinary.riscv> SIM_FLAGS="+jtag_rbb_enable=1 --rbb-port=9823" run-binary``)
3. Launch OpenOCD
4. Connect to OpenOCD via GDB
5. Done!

Example Test Chip Bringup Communication
---------------------------------------

Intro to Typical Chipyard Test Chip
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Most, if not all, Chipyard configurations are tethered using TSI (over a serial link) and have access
to external memory through an AXI port (backing AXI memory).
The following image shows the DUT with these set of default signals:

.. image:: ../_static/images/default-chipyard-config-communication.png

In this setup, the serial-link is connected to the TSI/FESVR peripherals while the AXI port is connected
to a simulated AXI memory.
However, AXI ports tend to have many signals associated with them so instead of creating an AXI port off the DUT,

> one can send the memory transactions over the bi-directional serial-link (``TLSerdesser``) so that main

Reviewer 2: ```suggestion
one can send the memory transactions over the bi-directional serial-link (``TLSerdesser``) so that the main
```

> interface to the DUT is the serial-link (which as comparatively less signals than an AXI port).

Reviewer 2: ```suggestion
interface to the DUT is the serial-link (which has comparatively less signals than an AXI port).
```
